mpqp/core/instruction/gates/custom_gate.py

```diff
@@ -82,12 +83,55 @@ def to_other_language(
         self,
         language: Language = Language.QISKIT,
         qiskit_parameters: Optional[set["Parameter"]] = None,
+        qcircuit: Optional["QCircuit"] = None,
     ):
-        from qiskit.quantum_info.operators import Operator as QiskitOperator
-
-        if qiskit_parameters is None:
-            qiskit_parameters = set()
-        return QiskitOperator(self.matrix)
+        if language == Language.QISKIT:
+            from qiskit.quantum_info.operators import Operator as QiskitOperator
+
+            if qiskit_parameters is None:
+                qiskit_parameters = set()
+            return QiskitOperator(self.matrix)
+        elif language == Language.QASM2:
+            from mpqp.tools.circuit import replace_custom_gate
+            from qiskit.quantum_info.operators import Operator as QiskitOperator
+            from qiskit import QuantumCircuit, qasm2
+
+            if qcircuit:
+                nb_qubits = qcircuit.nb_qubits
+            else:
+                nb_qubits = len(self.targets)
+
+            qiskit_circ = QuantumCircuit(nb_qubits)
+
+            qiskit_circ.unitary(
+                QiskitOperator(self.matrix).to_instruction(),
+                list(reversed(self.targets)),  # dang qiskit qubits order
+                self.label,
+            )
+            filtered_qasm = ""
+            final_gphase = 0
+            for instruction in qiskit_circ.data:
```

Comment:
do you propose to get instructions from `qiskit_circ.data[0]` ? If that's the case, I think it's better to use a for

Reply:
maybe
```py
qiskit_gate = QiskitOperator(self.matrix).to_instruction()
qiskit_circ.unitary(
    qiskit_gate,
    list(reversed(self.targets)),  # dang qiskit qubits order
    self.label,
)
circuit, gphase = replace_custom_gate(qiskit_gate, nb_qubits)
...
```

Reply:
we need that the instructions are instantiated in the circuit. so we can't just pass `qiskit_gate`

Reply:
haaa, in that case indeed both your solution and Hamza's one are the best suited, on one hand you're is "cleaner" (no special case) but Hamza's is clearer (in reality we do not iterate, since we know there is only one item), I'll let you discuss with him what to chose.

Reply:
done